Hi, we also use gcc internally to build the compiler and were seeing the warnings you are trying to fix here, so I tried applying your fix on our repo and it seems to fail with gcc 9.4.0 as well for the same errors I'm seeing with my Windows build of your fix. Namely this error:
```
/home/dyung/src/git/merge/llvm/include/llvm/ADT/PagedVector.h:220:40: error: ‘size_t llvm::PagedVector<clang::QualType>::Size’ is private within this context
  220 |       assert(LHS.ElementIdx == LHS.PV->Size ||
      |                                ~~~~~~~~^~~~
In file included from /home/dyung/src/git/merge/clang/include/clang/Basic/SourceManager.h:46,
                 from /home/dyung/src/git/merge/clang/lib/Serialization/ASTReader.cpp:61:
/home/dyung/src/git/merge/llvm/include/llvm/ADT/PagedVector.h:46:10: note: declared private here
   46 |   size_t Size = 0;
      |          ^~~~
...
/home/dyung/src/git/merge/llvm/include/llvm/ADT/PagedVector.h:222:23: error: ‘llvm::SmallVector<clang::QualType*, 0> llvm::PagedVector<clang::QualType>::PageToDataPtrs’ is private within this context
  222 |               LHS.PV->PageToDataPtrs[LHS.ElementIdx / PageSize]));
      |               ~~~~~~~~^~~~~~~~~~~~~~
In file included from /home/dyung/src/git/merge/clang/include/clang/Basic/SourceManager.h:46,
                 from /home/dyung/src/git/merge/clang/lib/Serialization/ASTReader.cpp:61:
/home/dyung/src/git/merge/llvm/include/llvm/ADT/PagedVector.h:50:31: note: declared private here
   50 |   mutable SmallVector<T *, 0> PageToDataPtrs;
      |                               ^~~~~~~~~~~~~~
```

--- a/llvm/include/llvm/ADT/PagedVector.h
+++ b/llvm/include/llvm/ADT/PagedVector.h
@@ -215,8 +215,15 @@ public:
       return LHS.equals(RHS);
     }
 
+    [[nodiscard]] size_t getIndex() const { return ElementIdx; }
+
+    friend bool operator!=(const MaterializedIterator &LHS,
+                           const MaterializedIterator &RHS) {
+      return !(LHS == RHS);
+    }
+
   private:
-    void verify() {
+    void verify() const {
       assert(ElementIdx == PV->Size ||
              (ElementIdx < PV->Size &&
               PV->PageToDataPtrs[ElementIdx / PageSize]));
@@ -228,13 +235,6 @@ public:
       Other.verify();
       return ElementIdx == Other.ElementIdx;
     }
-
-    friend bool operator!=(const MaterializedIterator &LHS,
-                           const MaterializedIterator &RHS) {
-      return !(LHS == RHS);
-    }
-
-    [[nodiscard]] size_t getIndex() const { return ElementIdx; }
   };
 
   /// Iterators over the materialized elements of the vector.

--- a/llvm/include/llvm/ADT/PagedVector.h
+++ b/llvm/include/llvm/ADT/PagedVector.h
@@ -212,18 +212,21 @@ public:
     /// Equality operator.
     friend bool operator==(MaterializedIterator const &LHS,
                            MaterializedIterator const &RHS) {
-      assert(LHS.PV == RHS.PV);
-      // Make sure we are comparing either end iterators or iterators pointing
-      // to materialized elements.
-      // It should not be possible to build two iterators pointing to non
-      // materialized elements.
-      assert(LHS.ElementIdx == LHS.PV->Size ||
-             (LHS.ElementIdx < LHS.PV->Size &&
-              LHS.PV->PageToDataPtrs[LHS.ElementIdx / PageSize]));
-      assert(RHS.ElementIdx == RHS.PV->Size ||
-             (RHS.ElementIdx < RHS.PV->Size &&
-              RHS.PV->PageToDataPtrs[RHS.ElementIdx / PageSize]));
-      return LHS.ElementIdx == RHS.ElementIdx;
+      return LHS.equals(RHS);
+    }
+
+  private:
+    void verify() {
+      assert(ElementIdx == PV->Size ||
+             (ElementIdx < PV->Size &&
+              PV->PageToDataPtrs[ElementIdx / PageSize]));
+    }
+
+    bool equals(const MaterializedIterator &Other) const {
+      assert(PV == Other.PV);
+      verify();
+      Other.verify();
+      return ElementIdx == Other.ElementIdx;
     }
 
     friend bool operator!=(MaterializedIterator const &LHS,

--- a/llvm/include/llvm/ADT/PagedVector.h
+++ b/llvm/include/llvm/ADT/PagedVector.h
@@ -210,8 +210,8 @@ public:
     }
 
     /// Equality operator.
-    friend bool operator==(MaterializedIterator const &LHS,
-                           MaterializedIterator const &RHS) {
+    friend bool operator==(const MaterializedIterator &LHS,
+                           const MaterializedIterator &RHS) {
       return LHS.equals(RHS);
     }
 
@@ -229,8 +229,8 @@ public:
       return ElementIdx == Other.ElementIdx;
     }
 
-    friend bool operator!=(MaterializedIterator const &LHS,
-                           MaterializedIterator const &RHS) {
+    friend bool operator!=(const MaterializedIterator &LHS,
+                           const MaterializedIterator &RHS) {
       return !(LHS == RHS);
     }
 

